tracking/inc/track.h

```diff
@@ -15,28 +17,153 @@ enum TrackStatus { alive, dead };
+  float total_length{0.0};        // length since track start
+   
+  float energy_loss{0};           // primitive version of scoring
+  int   number_of_secondaries{0}; // primitive version of scoring
 
-  __device__ double uniform() { return rng_state.Rndm(); }
+  uint16_t    num_step{0}; // number of Steps -- for killing after MaX
+   
+  char        current_process{0}; // -1 for Navigation / -2 for Field Propagation
+  char        pdg{0};             // Large enough for e/g. Then uint16_t for p/n
 
-  __device__ void SwapStates()
+  __device__ __host__ double uniform() { return rng_state.Rndm(); }
+
+  __device__ __host__ int    charge() const  // charge for e- / e+ / gamma only
+  { constexpr char pdgElec= 11;
+    int chrg= (pdg== -pdgElec ? 1 : 0)+(pdg==  pdgElec ? -1 : 0 ); return chrg; }
+                                    
+  __device__ __host__ double mass() const // Rest mass for e- / e+ / gamma only
+  { constexpr char pdgGamma= 22; return  (pdg== pdgGamma) ? 0.0 : copcore::units::kElectronMassC2; } 
+
+  __host__ void print( int id = -1, bool verbose = false) const;
+   
+  __device__ __host__ void SwapStates()
   {
     auto state          = this->current_state;
     this->current_state = this->next_state;
     this->next_state    = state;
   }
 };
 
+struct NavigationStateBuffer {
+   unsigned int currentTouchIndex;
+   unsigned int nextTouchIndex;
+   int currentLevel;
+   int nextLevel;   
+};
+
+// inline
+__global__
+void GetNavStateIndices( const track           & trk_dev,         // Unified or device
+                         NavigationStateBuffer & navState_unif  // Return => Unified memory
+   )
+{
+   // Needed because type of NavStateIndex (in track) is vecgeom:L:cxx::NavStateIndex
+   // Solution for printing by Andrei Gheata.
+   //
+   // trk.current_state.IsOnBoundary();
+   // std::cout << " current: "; // <<    setw(3) << next_state.IsOnBoundary();
+   assert ( NavIndAddr(trk_dev.current_state) != nullptr && "Invalid Current state in track");
+   
+   auto currentLevel      = trk_dev.current_state.GetLevel();
+
+   navState_unif.currentTouchIndex = trk_dev.current_state.GetNavIndex(currentLevel);
+   
+   navState_unif.currentLevel      = currentLevel;
+   // auto currentPhysVolume = trk_dev.current_state.ValueAt(currentLevel);
+   
+   // std::cout << " next: "; // <<    setw(3) << next_state.IsOnBoundary();
+   assert ( trk_dev.next_state != nullptr && "Invalid Next state in track");
+   assert ( NavIndAddr(trk_dev.next_state) != nullptr && "Invalid Next state in track");
+
+   auto nextLevel = trk_dev.next_state.GetLevel();     // This crashes consistently - in gdb 2020.12.17-18
+   navState_unif.nextTouchIndex = trk_dev.next_state.GetNavIndex(nextLevel);
+
+   navState_unif.nextLevel   = nextLevel;
+      
+   // auto nextTouchVolume = trk_dev.next_state.ValueAt(nextLevel);   
+}
+
+#include <cstdio>
+#include <iomanip>
+
+// inline
+__host__
+void track::print( int extId , bool verbose ) const
```

Comment:
I'm still concerned about this debugging code in a global header file: Kernel launches using the angle brackets is not valid C++, so host code trying to include this file will probably not compile.

Reply:
I understand your concerns and share them.

I am open to suggestions how to organise it better, while keeping the functionality.  Can we resolve this in a future PR, though?

Reply:
Yes, ok. As long as it builds right now, we can iterate on this down the road.

Reply:
We agreed to solve this in the future - thanks!